### DISCUSSION TITLE
configuring pyright-langserver settings

### DISCUSSION THREAD

**Discussion by StefPac** [joaotavora/eglot#1289] [Q&A] [answered]
Hello all,

I'm sure the answer must be right in front of my eyes but I can't seem to be able to have `pyright-langserver` accept settings such as `python.pythonPath` from `eglot`. Setting the variable `lsp-pyright-python-executable-cmd` in `lsp-pyright` and restarting the workspace accomplishes that, but I have not been able to reproduce the same with eglot. 

To accomplish that I have created a `.dir-locals.el` in the root of the project with contents:

```
(( nil . ((eglot-workspace-configuration
       . (:python
          (:pythonPath "path/to/my/interpreter"
           :analysis
             (:extraPaths ["foo" "bar"])))))))
```

I see the log entry:

```
(:jsonrpc "2.0" :method "workspace/didChangeConfiguration" :params
          (:settings
           (:python
            (:pythonPath "path/to/my/interpreter" :analysis
                         (:extraPaths
                          ["foo" "bar"])))))
```

But `pyright-langserv` does not seem to output the `Setting pythonPath for service ...` as it does when invoked from `pyright`.

Tested with directory structure:

```
proj_root \
            +.dir_locals.el
            + test.py
            + my_interpreter_executable
```
and command `emacs -q proj_root/test.py`

emacs version: `GNU Emacs 29.1 (build 2, x86_64-pc-linux-gnu) of 2023-08-26`
OS: Linux, Ubuntu, LTS 

Any help would be very welcome.

**Comment by joaotavora** [comment]
Hello.  Thanks to your self-contained recipe, I think I can fix your problem or at least advance this investigation:

I looked at the Eglot manual: https://joaotavora.github.io/eglot/#Advanced-server-configuration

And I used this `.dir-locals.el` file (full contents, note that your file is garbled as you pasted it):
```lisp
((python-mode
  . ((eglot-workspace-configuration
      . (:python
         (:pythonPath "path/to/my/interpreter"
                      :analysis
                      (:extraPaths ["foo" "bar"]))))
     (indent-tabs-mode . nil))))
```

After I setup the trivial project and do `C-u M-x eglot RET pyright-langserver --verbose --stdio`, I see this pop up in the Eglot events buffer eventually:

```
[server-notification] Thu Sep 14 02:26:09 2023:
(:jsonrpc "2.0" :method "window/logMessage" :params
          (:type 3 :message
                 "Setting pythonPath for service \"~/tmp/test-eglot-1289/\": \"/home/capitaomorte/tmp/test-eglot-1289/path/to/my/interpreter\""))
```

This is what you want right?

**Reply by StefPac** [reply]
Sorry for some reason the first line of the snippet stayed on the opening ``` line. I have fixed it. 

I don't fully understand the difference between my `.dir_locals.el` and yours but will try yours. I also didn't use the `--verbose` flag for `pyright-langserver` so I will try that first and see if that was why I could not see that it picked up the `pythonPath` anyway (I don't think so though because I see it not having certain python packages in the path when working on my code). 

Will report back shortly. Thanks much for the feedback.

**Reply by joaotavora** [reply]
I also don't think `--verbose` has anything to do with it.   I was just trying to get `stderr` output, but I guess pyright doesn't do that.

If it helps, my `pyright --version` reports `pyright 1.1.326`.  I'm using Eglot on latest Emacs master, but I think that doesn't make any difference.

I've also tried a variation of `.dir-locals.el` with `nil` instead of `python-mode`.  Same successful outcome.

**Reply by StefPac** [reply]
I'm running `pyright 1.1.325`  and need to figure out how to get it up to the latest. Will experiment and report back, thanks again a lot for the comments and for the great software you brought to the community.

**Reply by StefPac** [reply]
Adding more details for reference:
```
eglot: 1.15 
Python 3.10.11
pyright 1.1.327  # upgraded from 1.1.325
```

With either version of the `.dir_locals.el`I don't get that  `"Setting pythonPath for service ... "` message. 
Attached is the JSON-RPC [log from eglot](https://github.com/joaotavora/eglot/files/12612355/eglot.log)

**Reply by StefPac** [reply, marked as answer]
Had some time today and tested with emacs `master` (same build as my 29.1 one), which was the only difference that jumped at me between our setups. And surprise surprise... 🥁 

it worked. 

I wonder if https://github.com/joaotavora/eglot/pull/1281 was the fix for this

**Reply by StefPac** [reply]
For those following from home, it's good to notice that `pyright-1.1.327` has a regression from `1.1.326` creating problems with symlinks which creates confounding observations when debugging. 

@joaotavora I have not found the time to identify the exact commit that resolves this issue, but my problem is resolved. Unless there's something I can easily test for you, thank you for your help and keep up the great work.

**Reply by joaotavora** [reply]
> I wonder if https://github.com/joaotavora/eglot/pull/1281 was the fix for this

It might have been.  Note that that "fix" in Eglot is actually appeasing buggy server behaviour.

But glad to know that it works now.
